pkg/services/ngalert/migration/persist.go

```diff
@@ -0,0 +1,182 @@
+		return nil
+	}
+
+	for _, pair := range delta.ChannelsToAdd {
+		amConfig.AddReceiver(pair.ContactPoint)
+		amConfig.AddRoute(pair.Route)
+	}
+
+	// Validate the alertmanager configuration produced, this gives a chance to catch bad configuration at migration time.
+	// Validation between legacy and unified alerting can be different (e.g. due to bug fixes) so this would fail the migration in that case.
+	if err := sync.validateAlertmanagerConfig(amConfig.Config); err != nil {
+		return fmt.Errorf("validate AlertmanagerConfig: %w", err)
+	}
+
+	sync.log.Info("Writing alertmanager config", "receivers", len(amConfig.Config.AlertmanagerConfig.Receivers), "routes", len(amConfig.Config.AlertmanagerConfig.Route.Routes))
+	if err := sync.migrationStore.SaveAlertmanagerConfiguration(ctx, sync.orgID, amConfig.Config); err != nil {
+		return fmt.Errorf("write AlertmanagerConfig: %w", err)
+	}
+
+	return nil
+}
+
+// handleAddRules persists the given add rule delta to the state and database.
+func (sync *Sync) handleAddRules(ctx context.Context, state *migrationStore.OrgMigrationState, delta StateDelta) error {
+	createdFolderUIDs := make(map[string]struct{})
+	if len(delta.DashboardsToAdd) > 0 {
+		for _, duToAdd := range delta.DashboardsToAdd {
+			if _, ok := createdFolderUIDs[duToAdd.NewFolderUID]; duToAdd.CreatedFolder && duToAdd.NewFolderUID != "" && !ok {
+				createdFolderUIDs[duToAdd.NewFolderUID] = struct{}{}
+				state.CreatedFolders = append(state.CreatedFolders, duToAdd.NewFolderUID)
+			}
+			rules := make([]models.AlertRule, 0)
+			for _, pair := range duToAdd.MigratedAlerts {
+				rules = append(rules, *pair.Rule)
+			}
+			if len(rules) > 0 {
+				sync.log.Info("Inserting migrated alert rules", "count", len(rules))
+				err := sync.migrationStore.InsertAlertRules(ctx, rules...)
+				if err != nil {
+					return fmt.Errorf("insert alert rules: %w", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// validateAlertmanagerConfig validates the alertmanager configuration produced by the migration against the receivers.
+func (sync *Sync) validateAlertmanagerConfig(config *apiModels.PostableUserConfig) error {
+	for _, r := range config.AlertmanagerConfig.Receivers {
+		for _, gr := range r.GrafanaManagedReceivers {
+			data, err := gr.Settings.MarshalJSON()
+			if err != nil {
+				return err
+			}
+			var (
+				cfg = &alertingNotify.GrafanaIntegrationConfig{
+					UID:                   gr.UID,
+					Name:                  gr.Name,
+					Type:                  gr.Type,
+					DisableResolveMessage: gr.DisableResolveMessage,
+					Settings:              data,
+					SecureSettings:        gr.SecureSettings,
+				}
+			)
+
+			_, err = alertingNotify.BuildReceiverConfiguration(context.Background(), &alertingNotify.APIReceiver{
+				GrafanaIntegrations: alertingNotify.GrafanaIntegrations{Integrations: []*alertingNotify.GrafanaIntegrationConfig{cfg}},
+			}, sync.getDecryptedValue)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
```

Comment:
This file contains pieces moved from other files. Will it be hard to move tests too?

Reply:
I agree, but it will make rebasing a bit crazy to do this right now. I'll improve this later on.